### PR TITLE
[spark-rapids] update for 2024/07

### PR DESCRIPTION
**Pull request by cjac**
* no reboots
* added dkms signing code
* added utility functions for checking OS
* cleaned up URL generation a bit
* using full path to bdconfig so that the script works from `sudo -i` prompt
* fixed error messages to indicate that secure boot with driver signing is available on 2.2
* cleaned up remove_old_backports
* clean apt cache before exit

### CODE REVIEW

**Comment by cjac** [comment]
manually tested on

✅ 2.2-rocky9 
✅ 2.1-rocky8
❓ 2.0-rocky8
❓ 2.2-ubuntu22
❓ 2.1-ubuntu20
❓ 2.0-ubuntu18
❓ 2.2-debian12
❓ 2.1-debian11
❓ 2.0-debian10

**Comment by cjac** [comment]
@nvliyuan - would you please review this work with me?

Reboots caused a bunch of problems, so I've modified the script to install without a reboot.

I am also picking up a bunch of shared functions from the gpu installer and using them to simplify the code.  I believe that I've maintained all intended logic paths, but would appreciate it if NV could take a look.

Driver signing is working for 2.2, but not before.  See custom-images:
https://github.com/GoogleCloudDataproc/custom-images/tree/master/examples/secure-boot

I've got to run out to dinner, but I'll be back shortly to check on the manual cluster builds.

**Comment by nvliyuan** [comment]
@cjac thanks for the script updates, since we will drop support spark3.1* from the next release(24.08), so we only need to test 2.2 and 2.1 images. Today I will take some time to manually test 2.1/2.0 ubuntu/debian.

**Comment by nvliyuan** [comment]
✅ 2.1-ubuntu20
✅ 2.1-debian11
×  2.2-debian12 ： failed with below error while enabling secure boot
```
+ [[ enabled == \e\n\a\b\l\e\d ]]
+ echo 'Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.'
Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.
```
×  2.2-ubuntu22 ： failed with below error while enabling secure boot
```
+ [[ enabled == \e\n\a\b\l\e\d ]]
+ echo 'Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.'
Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.
```

**Comment by cjac** [comment]
Good to know.  I imagine that a future release of Dataproc will target 24.04 LTS and that once Trixie (Debian 13) is released, we will no longer support Debian 10.  We would benefit from a standard operating procedure for obsoleting deprecated releases.  The problem is that it's often less effort for customers to create a cluster and stick with it than it is for them to have a mental model of rolling releases.  Hopefully we can take some of that burden from their shoulders by ensuring that upgrades occur smoothly.

**Comment by cjac** [comment]
> ✅ 2.1-ubuntu20 ✅ 2.1-debian11 × 2.2-debian12 ： failed with below error while enabling secure boot
> 
> ```
> + [[ enabled == \e\n\a\b\l\e\d ]]
> + echo 'Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.'
> Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.
> ```
> 
> × 2.2-ubuntu22 ： failed with below error while enabling secure boot
> 
> ```
> + [[ enabled == \e\n\a\b\l\e\d ]]
> + echo 'Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.'
> Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster.
> ```

Okay, I thought I changed that wording.  Were you able to generate a custom image using the instructions?  I would love to hear how that process was for you.

File: spark-rapids/spark-rapids.sh [hunk 1]
```diff
@@ -551,44 +723,39 @@ function upgrade_kernel() {
...
+  elif is_rocky ; then
+    if ! is_rocky8 && ! is_rocky9 ; then
+      echo "Error: The Rocky Linux version ($(os_version)) is not supported. Please use a compatible Rocky Linux version."
       exit 1
     fi
   fi
 
-  if [[ "${SECURE_BOOT}" == "enabled" ]]; then 
-    echo "Error: Secure Boot is enabled. Please disable Secure Boot while creating the cluster."
+  if [[ "${SECURE_BOOT}" == "enabled" && $(echo "${DATAPROC_IMAGE_VERSION} <= 2.1" | bc -l) == 1 ]]; then
```
Comment (cjac) [inline spark-rapids/spark-rapids.sh:748]: @nvliyuan - I changed the error message here.  You must be running and older version of the script.

**Comment by nvliyuan** [comment]
ok, seems I tested with the wrong scripts(IIRC there is a spark-rapids.sh~), let me check it again(with https://github.com/cjac/initialization-actions/raw/spark-rapids-20240725/spark-rapids/spark-rapids.sh).

**Comment by cjac** [comment]
> ok, seems I tested with the wrong scripts(IIRC there is a spark-rapids.sh~), let me check it again(with https://github.com/cjac/initialization-actions/raw/spark-rapids-20240725/spark-rapids/spark-rapids.sh).

Yes, you're right.  I accidentally added and committed one of my emacs temp files.  I've added *~ to the .gitignore in hopes of not doing that again.

**Comment by nvliyuan** [comment]
For ubuntu22 and debian12 images, while enable secure boot, the error message is printed as expected:
```
+ echo 'Secure boot is enabled, but no signing material provided.'
Secure boot is enabled, but no signing material provided.
+ echo 'Please either disable secure boot or provide signing material as per'
Please either disable secure boot or provide signing material as per
+ echo https://github.com/GoogleCloudDataproc/custom-images/tree/master/examples/secure-boot
```
2.1-ubuntu20 also works well.

But I hit an error while testing 2.1-debian11, any insight about this error:
```
Hit:14 https://deb.debian.org/debian bullseye-updates InRelease
Reading package lists...
E: The repository 'https://archive.debian.org/debian bullseye-backports Release' does not have a Release file.
W: Target Packages (main/binary-amd64/Packages) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
W: Target Packages (main/binary-all/Packages) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
W: Target Translations (main/i18n/Translation-en) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
+ sleep 5
+ (( i++ ))
+ (( i < 10 ))
+ eval 'apt-get --allow-releaseinfo-change update'
++ apt-get --allow-releaseinfo-change update
Hit:1 https://storage.googleapis.com/goog-dataproc-bigtop-repo-us-central1/2_1_deb11_20240714_132421-RC01 dataproc InRelease
Hit:2 https://download.docker.com/linux/debian bullseye InRelease
Hit:3 https://packages.cloud.google.com/apt google-cloud-logging-bullseye-all InRelease
Ign:4 https://archive.debian.org/debian bullseye-backports InRelease
Hit:5 https://storage.googleapis.com/dataproc-bigtop-repo/2_1_deb11_20240714_132421-RC01 dataproc InRelease
Err:6 https://archive.debian.org/debian bullseye-backports Release
  404  Not Found [IP: 151.101.66.132 443]
Hit:7 https://packages.cloud.google.com/apt google-cloud-monitoring-bullseye-all InRelease
Hit:8 https://packages.cloud.google.com/apt google-compute-engine-bullseye-stable InRelease
Get:9 https://packages.adoptium.net/artifactory/deb bullseye InRelease [7517 B]
Hit:10 https://packages.cloud.google.com/apt cloud-sdk-bullseye InRelease
Hit:11 https://repo.mysql.com/apt/debian bullseye InRelease
Hit:12 https://deb.debian.org/debian bullseye InRelease
Hit:13 https://deb.debian.org/debian-security bullseye-security InRelease
Hit:14 https://deb.debian.org/debian bullseye-updates InRelease
Reading package lists...
E: The repository 'https://archive.debian.org/debian bullseye-backports Release' does not have a Release file.
W: Target Packages (main/binary-amd64/Packages) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
W: Target Packages (main/binary-all/Packages) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
W: Target Translations (main/i18n/Translation-en) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
+ sleep 5
+ (( i++ ))
+ (( i < 10 ))
+ eval 'apt-get --allow-releaseinfo-change update'
++ apt-get --allow-releaseinfo-change update
Hit:1 https://storage.googleapis.com/goog-dataproc-bigtop-repo-us-central1/2_1_deb11_20240714_132421-RC01 dataproc InRelease
Hit:2 https://packages.cloud.google.com/apt google-cloud-logging-bullseye-all InRelease
Hit:3 https://download.docker.com/linux/debian bullseye InRelease
Ign:4 https://archive.debian.org/debian bullseye-backports InRelease
Err:5 https://archive.debian.org/debian bullseye-backports Release
  404  Not Found [IP: 151.101.194.132 443]
Hit:6 https://packages.cloud.google.com/apt google-cloud-monitoring-bullseye-all InRelease
Hit:7 https://storage.googleapis.com/dataproc-bigtop-repo/2_1_deb11_20240714_132421-RC01 dataproc InRelease
Hit:8 https://packages.cloud.google.com/apt google-compute-engine-bullseye-stable InRelease
Hit:9 https://packages.cloud.google.com/apt cloud-sdk-bullseye InRelease
Hit:10 https://deb.debian.org/debian bullseye InRelease
Hit:11 https://deb.debian.org/debian-security bullseye-security InRelease
Hit:12 https://deb.debian.org/debian bullseye-updates InRelease
Hit:13 https://repo.mysql.com/apt/debian bullseye InRelease
Hit:14 https://packages.adoptium.net/artifactory/deb bullseye InRelease
Reading package lists...
E: The repository 'https://archive.debian.org/debian bullseye-backports Release' does not have a Release file.
W: Target Packages (main/binary-amd64/Packages) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
W: Target Packages (main/binary-all/Packages) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
W: Target Translations (main/i18n/Translation-en) is configured multiple times in /etc/apt/sources.list:7 and /etc/apt/sources.list:8
+ sleep 5
+ (( i++ ))
+ (( i < 10 ))
+ return 1
```
here is my script to create the cluster:
```
export REGION=us-central1
export GCS_BUCKET=spark_etl_tpch
export CLUSTER_NAME=ub20-t4-yuali-cjac-debian21
export NUM_GPUS=2
export NUM_WORKERS=2

gcloud dataproc clusters create $CLUSTER_NAME  \
    --region=$REGION \
    --zone="us-central1-a" \
    --image-version=2.1-debian11 \
    --master-machine-type=n1-standard-8 \
    --num-workers=$NUM_WORKERS \
    --worker-accelerator=type=nvidia-tesla-t4,count=$NUM_GPUS \
    --initialization-actions gs://rapids-test/yuanli-tools-eventlog-temp/gcp-init/v2406/spark-rapids-2406-cjac.sh  \
     --worker-machine-type=n1-standard-8 \
    --bucket=$GCS_BUCKET \
    --enable-component-gateway \
    --no-shielded-secure-boot \
    --subnet=default \
    --max-idle 2h
```

**Comment by nvliyuan** [comment]
To summary based on my test:
2.1-ubuntu20 : worked well
2.2-ubuntu22(diable secureboot): worked well
2.1-ubuntu22(enable secureboot): worked well, the error message printed as expected
2.1-debian11: hit error : https://github.com/GoogleCloudDataproc/initialization-actions/pull/1204#issuecomment-2251895882
2.2-debian12(diable secureboot): worked well
2.2-debian12(enable secureboot): worked well, the error message printed as expected

**Comment by nvliyuan** [comment]
the codes LGTM

**Comment by cjac** [comment]
The 2.1-debian11 issue you were seeing was fixed in a later commit.  Should be good now.  I'm having a hard time with signed 2.2-ubuntu22 images because of ubuntu's implementation of secure boot.  Trying to work around it...

https://github.com/GoogleCloudDataproc/initialization-actions/pull/1204/commits/a7896c6b5af5e1b2a12ea90a358f1c745449ea15#diff-d27cbf9feaa70fca8ee2730cd690de6a4ee10857b11f3ba7fda80cdcff0b00d1L772

**Comment by cjac** [comment]
Although the last commit does fix the signing problem, I think we can make a more surgical change and continue to use the code NV provided instead of using my work-around...

**Comment by nvliyuan** [comment]
> 2.1-debian11: hit error : [#1204 (comment)](https://github.com/GoogleCloudDataproc/initialization-actions/pull/1204#issuecomment-2251895882)

worked

**Review by prince-cs — APPROVED** [review]

**Pull request merged**